### ISSUE TITLE
LazyList.take is not lazy

### ISSUE THREAD

**Issue by 01mf02** [ocaml-batteries-team/batteries-included#688]
The following program prints 5 zeros, and I expected it to print a zero every second.
However, because `LazyList.take` is not lazy, it will wait for 5 seconds, then print all zeros at once.

~~~ ocaml
open Batteries
let _ =
  LazyList.from (fun () -> Unix.sleep 1; 0) |>
  LazyList.take 5 |> (* remove this line to get lazy behaviour *)
  LazyList.iter (Printf.printf "%d%!")
~~~

I believe that the intended behaviour of LazyList.take should be more like:

~~~ ocaml
let ltake n l =
  let rec aux i l =
    if i = 0 then LazyList.nil
    else match LazyList.get l with
      Some (x, rest) -> lazy (LazyList.Cons (x, aux (i-1) rest))
    | None -> raise (LazyList.Invalid_index n)
  in if n < 0 then raise (LazyList.Invalid_index n) else aux n l
~~~

Replacing `LazyList.take` in the example above with `ltake` is not giving quite the right result yet, because it waits for *two* seconds (instead of one) before showing the first element, but at least it is better than waiting five seconds. Can somebody tell me how I can improve this further?

**Comment by gasche** [comment]
I think that there are two reasons why `take` is eager:

1. from an implementation point of view, it is implemented in terms of `split_at : int -> 'a t -> 'a t * 'a t` which is naturally implemented in a single traversal that is strict on the first segment of the returned list, and lazy in the second one. (Having a lazier split_at could be nice but challenging implementation-wise, so maybe not using split-at for take would be better)

2. from an interface point of view, the problem with lazy `take` is that it is a partial function whose failure would also occur lazily, after the list elements have been enumerated. Another way to describe the design tension is that if the interface was the "pure" `take : int -> 'a t -> 'a t option`, we could not make it lazy.

I think these two points suggest that there is something natural about `take` being strict. I understand that it is surprising, and that the strict interface should advertize its strictness through the type system by returning an `'a list` instead of a `'a t`. So maybe we should have:

```
val split_at_list_rev : int -> 'a t * 'a list * 'a t (* returns the left segment reversed; maybe only an internal function *)
val take_list : int -> 'a t -> 'a list (* strict *)
val take : int -> 'a t -> 'a t (* lazy *)
```

**Comment by 01mf02** [comment]
Yes, I think that would be a good solution!

**Comment by c-cube** [comment]
@gasche I don't get it. It's much more natural for `take` to be lazy, and it's [quite simple](https://github.com/c-cube/ocaml-containers/blob/8cd06599aea41e1caa7c291d7a07fec3ba108825/src/iter/CCLazy_list.ml#L56) imho to implement in a lazy way. Also, https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatLazyList.html doesn't mention any possible failure of `take`. Did I miss something?

**Comment by gasche** [comment]
I believe that the current implementation fails if there are not enough elements in the list. Now that you mention it, it is indeed wrong as the documentation clearly says "returns *up to* `n` elements of the list", so it should not fail in that case.

(Of course a lazy "take" is not difficult to implement, I was mentioning that a lazy "split_at" is harder. If you are looking for a programming challenge or have already implemented it, feel free to propose an implementation to include to replace the current one that would move to `split_at_list_rev`).

**Comment by c-cube** [comment]
I think `take` should be implemented lazily and independently of any `split_at` function. On the other hand, `split_at` should probably have the signature `split_at: int -> 'a t -> 'a list * 'a t`… And sorry, I lack the motivation to make an efficient lazy `split_at` tonight (not sure it's even really possible).

**Comment by c-cube** [comment]
Ok, I think I actually have something that should be relatively efficient (read: no crazy memory consumption, unless the `drop` part is evaluated first). I think it's still better and simpler to write `take` and `drop` separately and make `split_at n l = take n l, drop n l`, but well...

```ocaml

module L = struct
  type 'a t = 'a node lazy_t
  and 'a node =
    | Nil
    | Cons of 'a * 'a t
end

let rec of_list l = lazy (
  match l with
    | [] -> L.Nil
    | x :: tail -> L.Cons (x, of_list tail)
)

type 'a split_at_state =
  | St_taking of { taken: int; remaining: int; l: 'a L.t }
  | St_eager of 'a list (* the rest is evaluated *)

let split_at n l =
  let st = ref (St_taking {taken=0; remaining=n; l}) in
  let rec get_take () = lazy (
    match !st with
      | St_taking {remaining=0; _} -> L.Nil
      | St_taking {taken; remaining=n; l} ->
        begin match l with
          | lazy L.Nil -> L.Nil
          | lazy (L.Cons (x, tail)) ->
            st := St_taking {taken=taken+1; remaining=n-1; l=tail};
            L.Cons (x, get_take())
        end
      | St_eager [] -> L.Nil
      | St_eager (x::tail) ->
        L.Cons (x, of_list tail)
  ) in
  let get_drop = lazy (
    match !st with
      | St_eager _ -> assert false
      | St_taking {remaining; l; _} ->
        let rec drop_ n acc l = match n, l with
          | 0, _ ->
            st := St_eager (List.rev acc);
            Lazy.force l
          | _, lazy L.Nil ->
            st := St_eager (List.rev acc);
            L.Nil
          | _, lazy (L.Cons (x, tail)) ->
            drop_ (n-1) (x::acc) tail
        in
        let res = drop_ remaining [] l in
        res
  ) in
  get_take(), get_drop

(* test generator *)
let make_test() =
  let rec g n = lazy (
    if n=5 then L.Nil
    else (
      Printf.printf "generate %d\n%!" n;
      Unix.sleep 1;
      L.Cons (n, g (n+1))
    )
  ) in
  g 0

let get = function
  | lazy L.Nil -> None
  | lazy (L.Cons (x, tail)) -> Some (x,tail)

let rec iter_test l: unit = match l with
  | lazy L.Nil -> ()
  | lazy (L.Cons (x, tail)) ->
    Printf.printf "obtained %d\n%!" x;
    iter_test tail

let () =
  print_endline "simple iteration";
  iter_test (make_test());
  print_endline "take 3, then drop 2";
  let t1, t2 = make_test () |> split_at 3 in
  iter_test t1;
  iter_test t2;
  print_endline "take 1, then drop 2, then take 2";
  let t1, t2 = make_test () |> split_at 3 in
  ignore (get t1);
  iter_test t2;
  iter_test t1;
  ()
```

**Comment by gasche** [comment]
@c-cube what do you think of the following:

```ocaml
let split_at n li =
  let last_n = ref n in
  let last_li = ref li in
  let rec take n li =
    last_n := n;
    last_li := li;
    if n = 0 then lazy L.Nil
    else
      lazy
        (match (Lazy.force li) with
         | L.Nil -> L.Nil
         | L.Cons (x, xs) -> L.Cons (x, take (n - 1) xs))
  in
  take n li, lazy (Lazy.force (drop !last_n !last_li))
```

(Note that for a Seq-like structure one would also have to flow information back from `drop` to `take`, so that `take` does not redo the computation of nodes forced by `drop` if called after, but here lazyness takes care of that for us.)

**Comment by c-cube** [comment]
It looks ok and simpler, indeed. Did you test it with the interleavings in my snippet?

**Comment by gasche** [comment]
Yes, I think it has the same behavior as your code (as far as I could see).

**Comment by c-cube** [comment]
Then I suppose it's a good fix ^^

**Comment by 01mf02** [comment]
If the `split_at` version proposed by @gasche is going to replace the current one in batteries,
note that this version changes the semantics of the function, because it does not raise Invalid_index.

**Comment by 01mf02** [comment]
Any update on this? Is this going to be integrated into batteries?

**Comment by UnixJunkie** [comment]
I'm unsure if this should be closed or not.
